src/vcs/git/repo.rs

```diff
@@ -0,0 +1,388 @@
+// This file is part of radicle-surf
+// <https://github.com/radicle-dev/radicle-surf>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pub use crate::vcs::git::object::{Branch, BranchName, Commit, RevObject, Signature, TagName};
+use crate::{
+    file_system,
+    tree::*,
+    vcs,
+    vcs::{git::error::*, VCS},
+};
+use git2::{BranchType, Oid};
+use nonempty::NonEmpty;
+use std::{cmp::Ordering, convert::TryFrom, str};
+
+/// OrderedCommit is to allow for us to identify an ordering of commit history
+/// as we enumerate over a revwalk of commits, by assigning each commit an
+/// identifier.
+#[derive(Clone)]
+pub(super) struct OrderedCommit {
+    pub(super) id: usize,
+    pub(super) commit: Commit,
+}
+
+impl std::fmt::Debug for OrderedCommit {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "OrderedCommit {{ id: {}, commit: {} }}",
+            self.id, self.commit.id
+        )
+    }
+}
+
+impl OrderedCommit {
+    pub(super) fn compare_by_id(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id).reverse()
+    }
+}
+
+impl From<OrderedCommit> for Commit {
+    fn from(ordered_commit: OrderedCommit) -> Self {
+        ordered_commit.commit
+    }
+}
+
+/// A `History` that uses `git2::Commit` as the underlying artifact.
+pub type History = vcs::History<Commit>;
+
+/// Wrapper around the `git2`'s `git2::Repository` type.
+/// This is to to limit the functionality that we can do
+/// on the underlying object.
+pub struct Repository(pub(super) git2::Repository);
+
+/// A reference-only `Repository`. This means that we cannot mutate the
```

Comment:
The description doesn't quite match my understanding: `Send` only means that the value can be transferred across thread boundaries, _not_ that it can be shared among threads.

`git2::Repository` is `Send`, but to make it `Sync` (ie. shareable), we need to slap an `Arc<Mutex<git2::Repository>>`. If we wanted to get a `Browser` while holding the lock, we _should_ be able to something like this:

```rust
let locked_repo: MutexGuard<git2::Repository> = shared_repo.lock().unwrap();
let surf_repo = RepositoryRef(&locked_repo);
let bro = Browser::new(surf_repo);
// can do bro things while `locked_repo` is in scope
```

Reply:
Right, sorry. I should have read the docs first :grimacing:

Reply:
The question is if this really works. Perhaps we should add a test somewhere.

Reply:
Test sounds good :sweat_smile:

Reply:
https://github.com/radicle-dev/radicle-surf/pull/106/commits/ffa8f0747ef64f7c4d4fd7f75fefde906d141f4c